Update dependencies, add Scala Native support for circe integration, add jsoniter-scala integration

plokhotnyuk opened the pull request:
Close #6 by adding jsoniter-scala module

@ybasket @satabin Can I use code from this PR in a tutorial for jsoniter-scala?

plokhotnyuk committed:
Update dependencies, add Scala Native support for circe integration, add jsoniter-scala integration

ybasket commented:
@plokhotnyuk thank you for contributing! I unfortunately don't have the time right now to dive deep into how the codecs work, but it looks good at first glance and I'll try to find some time soon-ish if @satabin isn't faster.

As for using the code in a tutorial, no objections from my side, looking forward to read it. A link back to geo-scala would of course be appreciated :)

plokhotnyuk committed:
Update tlBaseVersion to 0.3.0

plokhotnyuk commented:
@ybasket Please approve workflow run, I hope the latest commit will fix CI build failures

ybasket commented:
> @ybasket Please approve the workflow run, I hope the latest commit will fix CI build failures

Done. Sorry, we probably should have updated after the last release (still learning the details of sbt-typelevel).

ybasket commented on build.sbt:
```diff
@@ -1,8 +1,8 @@
+val scala3 = "3.2.0"
 
-ThisBuild / tlBaseVersion := "0.2"
+ThisBuild / tlBaseVersion := "0.3.0"
```
```suggestion
ThisBuild / tlBaseVersion := "0.3"
```

ybasket committed:
Update build.sbt

plokhotnyuk committed:
Update result of `sbt githubWorkflowGenerate`

plokhotnyuk committed:
Fix compilation errors with Scala 3

plokhotnyuk committed:
Disable expected warnings

plokhotnyuk committed:
Clean up comments

plokhotnyuk commented:
Ups, it failed again... Now looking a way how to generate CI scripts with Java 11 for the `jsoniterScala` module...

plokhotnyuk committed:
A workaround to build Java 8 bytecode using Java 11

armanbilge commented on build.sbt:
```diff
@@ -25,6 +27,9 @@ lazy val commonSettings = Seq(
     "org.scalatestplus" %%% "scalacheck-1-16" % "3.2.13.0" % Test,
     "org.scalacheck" %%% "scalacheck" % "1.16.0" % Test
   ),
+  scalacOptions ++= Seq(
+    "-release 8"
+  ),
```
This is redundant to `ThisBuild / tlJdkRelease := Some(8)`

armanbilge commented on build.sbt:
```diff
@@ -16,6 +16,8 @@ ThisBuild / tlSonatypeUseLegacyHost := true
 
 ThisBuild / crossScalaVersions := Seq(scala212, scala213, scala3)
 ThisBuild / scalaVersion := scala213 // the default Scala
+ThisBuild / githubWorkflowJavaVersions += JavaSpec.temurin("11")
```
Probably, you want to `:= Seq(JavaSpec.temurin("8"))`. Since this will _add_ a JDK 11 job, instead of replacing the JDK 8 job.

armanbilge commented on build.sbt:
```diff
@@ -16,12 +16,14 @@ ThisBuild / tlSonatypeUseLegacyHost := true
 
 ThisBuild / crossScalaVersions := Seq(scala212, scala213, scala3)
 ThisBuild / scalaVersion := scala213 // the default Scala
+ThisBuild / githubWorkflowJavaVersions := Seq(JavaSpec.temurin("8"))
```
```suggestion
ThisBuild / githubWorkflowJavaVersions := Seq(JavaSpec.temurin("11"))
```

plokhotnyuk commented:
@armanbilge Thank you a lot! It looks like a clear solution (not a workaround) isn't it?

armanbilge commented:
Yes, I would consider it to be a solution. In fact there is a proposal to make a configuration like this the default for the next major release of sbt-typelevel.
- https://github.com/typelevel/sbt-typelevel/issues/215

Commit:
Run githubWorkflowGenerate to reflect JDK versions

plokhotnyuk commented:
Thank you, @ybasket! I missed to run the `githubWorkflowGenerate` task again

armanbilge commented:
Btw, I think this can be removed:
https://github.com/gnieh/geo-scala/blob/876e75ca8c5766944449eaa54ad4a90cbba01c78/build.sbt#L30-L32

armanbilge commented:
Note: you should _never_ set `mimaPreviousArtifacts := Set.empty`. This is wrong. You always fall into one of two scenarios:
1. You are not publishing that module. So `.enablePlugins(NoPublishPlugin)`.
2. It is a brand new artifact, not yet published. So set
  ```scala
    tlVersionIntroduced := Map("2.12" -> "1.2.3", "2.13" -> "1.2.3", "3" -> "1.2.3")
   ```
   Assuming that `1.2.3` is the first version it will be published in.

plokhotnyuk commented:
@armanbilge IMHO just removing of `mimaPreviousArtifacts := Set.empty` doesn't enable checking of binary compatibility...

Have the `sbt-typelevel` plugin something to replace the following workaround?
```
  mimaPreviousArtifacts := {
    import scala.sys.process._

    val oldVersion = "git describe --abbrev=0".!!.trim.replaceAll("^v", "")

    def isCheckingRequired: Boolean = {
      val Array(newMajor, _, _) = version.value.split('.')
      val Array(oldMajor, _, _) = oldVersion.split('.')
      newMajor == oldMajor
    }

    if (isCheckingRequired) Set(organization.value %% moduleName.value % oldVersion)
    else Set()
  }
```

armanbilge commented:
Yes, sbt-typelevel automatically determines the previous artifacts for MiMa binary-comaptibility checking. You can read more about it in the documentation.

https://typelevel.org/sbt-typelevel/faq.html#what-is-a-base-version-anyway

plokhotnyuk commented:
I rolled back and force pushed updating of docs but the CI build still fails in the static code analysis part...

plokhotnyuk commented:
What is Codacy Static Code Analysis? How to open logs or results of its checks?

plokhotnyuk committed:
Code and tests clean up

plokhotnyuk committed:
Add missing check for required fields

plokhotnyuk committed:
Fix formatting

plokhotnyuk committed:
Update docs

plokhotnyuk committed:
Clean up build

plokhotnyuk committed:
Add missing checking for `}` when parsing `Geometry` types

plokhotnyuk commented:
@ybasket @satabin Should I fix ["issues" reported by Codacy Static Code Analysis?](https://app.codacy.com/gh/gnieh/geo-scala/pullRequest?prid=10200627)

Some of them like in `README.md` are not fixable, others will lead to less readable and more error prone code...

satabin commented:
Hi @plokhotnyuk You can disregard Codacy, their profile is not tuned yet for our usage.

satabin commented on jsoniter-scala/src/main/scala/com/free2move/geoscala/jsoniter_scala.scala:
```diff
@@ -0,0 +1,156 @@
+import com.github.plokhotnyuk.jsoniter_scala.macros._
+import scala.annotation.nowarn
+
+object jsoniter_scala {
```
do we need the `_scala` in the package name?

plokhotnyuk replied:
Yes, [jsoniter](https://jsoniter.com/) is quite a different beast of software, and jsoniter-scala doesn't depend on it for ages

plokhotnyuk committed:
Update Scala 2.13.x to 2.13.9

plokhotnyuk committed:
Add missing tests for serialization and fix codecs to pass them

plokhotnyuk commented:
@ybasket @satabin This PR is ready for final review. 

I propose make other improvements for binary compatibility checks and replacement of `List` by `ArraySeq` in subsequent PRs

plokhotnyuk committed:
Fix formatting

plokhotnyuk commented:
@ybasket @satabin This PR is ready for final review. 

I propose make other improvements for binary compatibility checks and replacement of `List` by `ArraySeq` in subsequent PRs

plokhotnyuk committed:
Fix compilation error with Scala 2.12.x

plokhotnyuk committed:
Fix Scala 3 compiler warnings

plokhotnyuk commented:
@armanbilge I have tried `sbt +mimaReportBinaryIssues` without ` mimaPreviousArtifacts := Set.empty` line and it doesn't work. 

All modules are still unchecked due to empty `mimaPreviousArtifacts`.

We should set it explicitly, isn't it?

satabin commented:
Is the base version correctly set?

armanbilge commented:
@plokhotnyuk yes, that is the correct, expected behavior. Because you have bumped the base version from `0.2` -> `0.3` you have signified that you are making a breaking change, and therefore sbt-typelevel stops checking against previous versions.

You should not update the base version if you want to continue checking against previous versions.

I highly recommend you read the material that I linked you to before :) it explains this in detail, with examples.

https://typelevel.org/sbt-typelevel/faq.html#how-do-i-introduce-breaking-changes-intended-for-my-next-version
https://typelevel.org/sbt-typelevel/faq.html#what-is-a-base-version-anyway

plokhotnyuk commented:
@armanbilge I don't understand how all the magic should work... Locally when I revert the base version back to `0.2` it doesn't validate any module.

armanbilge commented:
Have you pulled the tags to your local clone? Otherwise it will be unable to detect previous versions locally (but tags are always present in CI).

armanbilge commented:
Btw, I see that v0.3.0 is already released. So now I am confused what is happening in your PR.
https://github.com/gnieh/geo-scala/releases/tag/v0.3.0

armanbilge commented:
For the record: if I leave the `tlBaseVersion := "0.3"` and I remove the `mimaPreviousArtifacts := Set.empty`, then it is all configured correctly.

```
sbt:geo-scala> show mimaPreviousArtifacts
[info] jsoniterScalaJVM / mimaPreviousArtifacts
[info]  Set(org.gnieh:geo-scala-jsoniter-scala:0.3.0 (e:info.versionScheme=early-semver))
[info] coreJS / mimaPreviousArtifacts
[info]  Set(org.gnieh:geo-scala-core:0.3.0 (e:info.versionScheme=early-semver))
[info] circeNative / mimaPreviousArtifacts
[info]  Set(org.gnieh:geo-scala-circe:0.3.0 (e:info.versionScheme=early-semver))
[info] jsoniterScalaJS / mimaPreviousArtifacts
[info]  Set(org.gnieh:geo-scala-jsoniter-scala:0.3.0 (e:info.versionScheme=early-semver))
[info] jsoniterScalaNative / mimaPreviousArtifacts
[info]  Set(org.gnieh:geo-scala-jsoniter-scala:0.3.0 (e:info.versionScheme=early-semver))
[info] circeJVM / mimaPreviousArtifacts
[info]  Set(org.gnieh:geo-scala-circe:0.3.0 (e:info.versionScheme=early-semver))
[info] coreNative / mimaPreviousArtifacts
[info]  Set(org.gnieh:geo-scala-core:0.3.0 (e:info.versionScheme=early-semver))
[info] polylineJVM / mimaPreviousArtifacts
[info]  Set(org.gnieh:geo-scala-polyline:0.3.0 (e:info.versionScheme=early-semver))
[info] circeJS / mimaPreviousArtifacts
[info]  Set(org.gnieh:geo-scala-circe:0.3.0 (e:info.versionScheme=early-semver))
[info] polylineJS / mimaPreviousArtifacts
[info]  Set(org.gnieh:geo-scala-polyline:0.3.0 (e:info.versionScheme=early-semver))
[info] coreJVM / mimaPreviousArtifacts
[info]  Set(org.gnieh:geo-scala-core:0.3.0 (e:info.versionScheme=early-semver))
[info] polylineNative / mimaPreviousArtifacts
[info]  Set(org.gnieh:geo-scala-polyline:0.3.0 (e:info.versionScheme=early-semver))
[info] mimaPreviousArtifacts
[info]  Set()
sbt:geo-scala> 
```

plokhotnyuk commented:
It doesn't work for me... Have I missed to run some sbt task?
```
andriy@notebook:~/Projects/com/github/plokhotnyuk/geo-scala$ sbt
[info] welcome to sbt 1.7.1 (Azul Systems, Inc. Java 17.0.2)
[info] loading global plugins from /home/andriy/.sbt/1.0/plugins
[info] loading settings for project geo-scala-build from plugins.sbt ...
[info] loading project definition from /home/andriy/Projects/com/github/plokhotnyuk/geo-scala/project
[info] loading settings for project root from build.sbt,publish.sbt ...
[info] resolving key references (15836 settings) ...
[info] set scmInfo to https://github.com/plokhotnyuk/geo-scala
[info] set current project to geo-scala (in build file:/home/andriy/Projects/com/github/plokhotnyuk/geo-scala/)
[info] sbt server started at local:///home/andriy/.sbt/1.0/server/0eabc375e6e88395df69/sock
[info] started sbt server
sbt:geo-scala> show mimaPreviousArtifacts
[info] coreJVM / mimaPreviousArtifacts
[info] 	Set()
[info] jsoniterScalaJVM / mimaPreviousArtifacts
[info] 	Set()
[info] circeJS / mimaPreviousArtifacts
[info] 	Set()
[info] coreJS / mimaPreviousArtifacts
[info] 	Set()
[info] polylineNative / mimaPreviousArtifacts
[info] 	Set()
[info] polylineJS / mimaPreviousArtifacts
[info] 	Set()
[info] circeNative / mimaPreviousArtifacts
[info] 	Set()
[info] jsoniterScalaJS / mimaPreviousArtifacts
[info] 	Set()
[info] circeJVM / mimaPreviousArtifacts
[info] 	Set()
[info] jsoniterScalaNative / mimaPreviousArtifacts
[info] 	Set()
[info] polylineJVM / mimaPreviousArtifacts
[info] 	Set()
[info] coreNative / mimaPreviousArtifacts
[info] 	Set()
[info] mimaPreviousArtifacts
[info] 	Set()
sbt:geo-scala> 
[info] shutting down sbt server
andriy@notebook:~/Projects/com/github/plokhotnyuk/geo-scala$ git diff
diff --git a/build.sbt b/build.sbt
index 59a5aa7..0432a9c 100644
--- a/build.sbt
+++ b/build.sbt
@@ -26,10 +26,7 @@ lazy val commonSettings = Seq(
     "org.scalatest" %%% "scalatest" % "3.2.13" % Test,
     "org.scalatestplus" %%% "scalacheck-1-16" % "3.2.13.0" % Test,
     "org.scalacheck" %%% "scalacheck" % "1.16.0" % Test
-  ),
-  // disable MiMa until we have proper version released on gnieh / all platforms
-  // once removed, the tlBaseVersion has to adjusted as well
-  mimaPreviousArtifacts := Set.empty
+  )
 )
 
 lazy val core = crossProject(JVMPlatform, JSPlatform, NativePlatform)
```

armanbilge commented:
@plokhotnyuk what is the output of `git tag` on your system?

plokhotnyuk commented:
@armanbilge Thanks for the tip! There were no tags in the local clone of my fork until I fetched them from the upstream:
```
andriy@notebook:~/Projects/com/github/plokhotnyuk/geo-scala$ git tag
andriy@notebook:~/Projects/com/github/plokhotnyuk/geo-scala$ git remote add upstream https://github.com/gnieh/geo-scala
andriy@notebook:~/Projects/com/github/plokhotnyuk/geo-scala$ git remote -v
origin	git@github.com:plokhotnyuk/geo-scala.git (fetch)
origin	git@github.com:plokhotnyuk/geo-scala.git (push)
upstream	https://github.com/gnieh/geo-scala (fetch)
upstream	https://github.com/gnieh/geo-scala (push)
andriy@notebook:~/Projects/com/github/plokhotnyuk/geo-scala$ git fetch upstream
remote: Enumerating objects: 51, done.
remote: Counting objects: 100% (51/51), done.
remote: Compressing objects: 100% (9/9), done.
remote: Total 36 (delta 9), reused 35 (delta 9), pack-reused 0
Unpacking objects: 100% (36/36), 5.09 KiB | 473.00 KiB/s, done.
From https://github.com/gnieh/geo-scala
 * [new branch]      feature/refresh-versions             -> upstream/feature/refresh-versions
 * [new branch]      fix-polyline-on-small-negative-delta -> upstream/fix-polyline-on-small-negative-delta
 * [new branch]      geohash                              -> upstream/geohash
 * [new branch]      main                                 -> upstream/main
 * [new branch]      switch-to-sbt-typelevel              -> upstream/switch-to-sbt-typelevel
 * [new branch]      upgrade-to-2.13                      -> upstream/upgrade-to-2.13
 * [new tag]         v0.1.0                               -> v0.1.0
 * [new tag]         v0.1.0-2023c94f                      -> v0.1.0-2023c94f
 * [new tag]         v0.1.0-907d7c6d                      -> v0.1.0-907d7c6d
 * [new tag]         v0.1.0-fb77b3a3                      -> v0.1.0-fb77b3a3
 * [new tag]         v0.1.1                               -> v0.1.1
 * [new tag]         v0.1.2                               -> v0.1.2
 * [new tag]         v0.2.0                               -> v0.2.0
 * [new tag]         v0.3.0                               -> v0.3.0
andriy@notebook:~/Projects/com/github/plokhotnyuk/geo-scala$ git tag
v0.1.0
v0.1.0-2023c94f
v0.1.0-907d7c6d
v0.1.0-fb77b3a3
v0.1.1
v0.1.2
v0.2.0
v0.3.0
```

plokhotnyuk committed:
Bump the base version and turn on binary compatibility checks

plokhotnyuk commented:
I bumped the base version due to changes in the circe module that replaced `io.circe.Encoder#AsObject` by `io.circe.Encoder` in some pubic methods which were required for passing added serialization assertions in tests.

Commit:
Fix Scala version in workflows

Commit:
Downgrade to Scala 2.13.8

ybasket commented:
https://github.com/scala/scala/pull/9982 produces some warnings while compiling that are then turned into errors 😢 

I downgraded to Scala 2.13.8 to avoid this. I didn't see anything in the release notes of 2.13.9 we would profit from…

ybasket commented:
@plokhotnyuk thank you for contributing this and your patience with the CI 👍. Would be interested to read that blog article, maybe you throw a link here once ready? 

And also thank you 👍, @armanbilge!

I'll release as 0.4.0.
